ICM45686 AN-000364 FIFO Fix

anthony289478 opened the pull request:
When running the icm45686 on current main for several hours, occasionally I hit this assert,
https://github.com/zephyrproject-rtos/zephyr/blob/3e7a941094b4b343345a35ab1477d87b7a49dafa/drivers/sensor/tdk/icm45686/icm45686_decoder.c#L492-L496

When inspecting the value of the header, it is 0xff which is invalid and lead me to this comment in the official TDK HAL driver.
[zephyrproject-rtos/hal_tdk/icm456xx/imu/inv_imu_driver_advanced.c](https://github.com/zephyrproject-rtos/hal_tdk/blob/f2b4a52a9d25be4fc244d3baac86c21cecde5bed/icm456xx/imu/inv_imu_driver_advanced.c#L856-L862)

Since the current icm45686 driver implementation reads the fixed watermark threshold number of frames, and in cases where the watermark is equal to 1, the workaround is the watermark threshold set to M + 1 frames and M frames are read out on each threshold trigger.

anthony289478 committed:
sensor: icm45686: Print unsupported fifo packet on assert

When an unsupported fifo packet triggers an assert, it is
helpful to see the packet as a hex value to help when debugging.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

anthony289478 committed:
sensor: icm45686: Implement AN-000364 fix for fifo corruption

When operating in streaming mode, a FIFO empty event, which
is caused by host reading the last byte of the last FIFO frame,
can cause FIFO data corruption.

During subsequent reads to the FIFO, the first  frame that arrives
after the empty condition will be corrupted. Once the issue occurs,
the internal state cannot recover and the FIFO must be flushed in
bypass mode to clear the corrupted state.

The current workaround from TDK is to read M-1 frames when M frames
are reported by fifo_count. Since M is fixed by the fifo_watermark
DT parameter, and in cases where fifo_watermark == 1, the watermark
trigger threshold is set to M + 1 frames and M frames are read out
during a watermark threshold event.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

ubieda commented on drivers/sensor/tdk/icm45686/icm45686_stream.c:
```diff
@@ -470,8 +470,16 @@ void icm45686_stream_submit(const struct device *dev,
 
 		if (data->stream.settings.enabled.fifo_ths ||
 		    data->stream.settings.enabled.fifo_full) {
+			/** AN-000364: When operating in FIFO streaming mode, if FIFO threshold
```
Thanks for the reference.

anthony289478 committed:
sensor: icm45686: Swap ASSERT to CHECKIF in fifo decode

Change the __ASSERT on unsupported fifo packet header to
CHECKIF so the user can choose how to handle the invalid
packet.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

anthony289478 committed:
sensor: icm45686: Use explicit VLA for fifo encoded data

The size of icm45686_encoded_data fifo_payload is guaranteed
to span the full number of frames read as it was allocated during
icm45686_event_handler() buf_len_required size.

This update suppresses the static analysis warning of
out of bounds memory access at index 1 by explicitly declaring
fifo_payload to be a VLA at the end of icm45686_encoded_data.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

yperess commented:
> I don't mean to veer off the PR topic, but I wonder if we should add a `size_t len` next to the `const uint8_t *buffer` on the Decoder Decode API: https://github.com/zephyrproject-rtos/zephyr/blob/052ded1682a9fd23728be8436dc8ded12c2fa9e0/include/zephyr/drivers/sensor.h#L485
> 
> Feels like, even though this is typically encoded in the data, it'd be safer to not entirely rely on the data when deciding how far to go pass the edata address (e.g: a bunch of 0xFF's in a corrupted buffer).
> 
> What do you think? @teburd @MaureenHelm @yperess

I'll need to go back and double check, but I think there was a reason for that. I seem to recall that it was hard or useless to guarantee the size at the decode phase. I'm not opposed to it though.
